tfx_bsl/beam/run_inference.py

```diff
@@ -117,19 +126,119 @@ def RunInferenceImpl(  # pylint: disable=invalid-name
+  these queries are grouped by model and inference runs in batches. If a
+  fixed_inference_spec_type is provided, this spec is used for all inference
+  requests which enables pre-configuring the model durign pipeline
+  construction.
+
+  Args:
+    queries: A PCollection containing QueryType tuples
+    fixed_inference_spec_type: an optional model inference endpoint. If
+      specified, this is "preloaded" during inference and models specified in
+      query tuples are ignored. This requires the InferenceSpecType to be known
+      at pipeline creation time.
+
+  Returns:
+    A PCollection containing prediction logs.
+
+  Raises:
+    ValueError; when operation is not supported.
+  """
+  batched_queries = queries | 'BatchQueries' >> _BatchQueries()
+  predictions = None
+
+  if fixed_inference_spec_type is None:
+    # operation type is determined at runtime
+    tagged = (
+      batched_queries | 'Tag by operation' >> beam.Map(
+        lambda batch: beam.pvalue.TaggedOutput(
+          _get_operation_type(batch[0]), batch)
+      )
+      .with_outputs(
+        OperationType.CLASSIFICATION,
+        OperationType.REGRESSION,
+        OperationType.PREDICTION,
+        OperationType.MULTIHEAD
+      )
+    )
+
+    predictions = [
+      tagged[OperationType.CLASSIFICATION] | 'Classify' >> _Classify(),
+      tagged[OperationType.REGRESSION] | 'Regress' >> _Regress(),
+      tagged[OperationType.PREDICTION] | 'Predict' >> _Predict(),
+      tagged[OperationType.MULTIHEAD] | 'MultiInference' >> _MultiInference()
+    ] | beam.Flatten()
   else:
-    raise ValueError('Unsupported operation_type %s' % operation_type)
+    # operation type is determined at pipeline construction time
+    operation_type = _get_operation_type(fixed_inference_spec_type)
+
+    if operation_type == OperationType.CLASSIFICATION:
+      predictions = batched_queries | 'Classify' >> _Classify(
+        fixed_inference_spec_type=fixed_inference_spec_type)
+    elif operation_type == OperationType.REGRESSION:
+      predictions = batched_queries | 'Regress' >> _Regress(
+        fixed_inference_spec_type=fixed_inference_spec_type)
+    elif operation_type == OperationType.PREDICTION:
+      predictions = batched_queries | 'Predict' >> _Predict(
+        fixed_inference_spec_type=fixed_inference_spec_type)
+    elif operation_type == OperationType.MULTIHEAD:
+      predictions = (batched_queries | 'MultiInference' >> _MultiInference(
+        fixed_inference_spec_type=fixed_inference_spec_type))
+    else:
+      raise ValueError('Unsupported operation_type %s' % operation_type)
+
+  return predictions
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(QueryType)
+@beam.typehints.with_output_types(_QueryBatchType)
+def _BatchQueries(queries: beam.pvalue.PCollection) -> beam.pvalue.PCollection:
+  """Groups queries into batches."""
+
+  def _add_key(query: QueryType) -> Tuple[bytes, QueryType]:
```

Comment:
Before streaming is enabled, the model will be the same during inference. Is this still needed?

Reply:
I've reduced `_BatchQueries` to the minimum required to operate on identical inference specs. #13 can introduce keyed batching and inference spec serialization